lib/resumption/resumable.go

```diff
@@ -0,0 +1,334 @@
+// Teleport
+// Copyright (C) 2023  Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resumption
+
+import (
+	"bufio"
+	"context"
+	"encoding/binary"
+	"io"
+	"net"
+	"sync/atomic"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+func newResumableConn(localAddr, remoteAddr net.Addr) *Conn {
+	r := &Conn{
+		managedConn: managedConn{
+			localAddr:  localAddr,
+			remoteAddr: remoteAddr,
+		},
+	}
+	r.cond.L = &r.mu
+	return r
+}
+
+// Conn is a [net.Conn] whose underlying transport can be closed and reopened,
+// to maintain the illusion of a perfect unbroken stream of bytes even if
+// network conditions would otherwise terminate a normal connection.
+type Conn struct {
+	managedConn
+
+	// attached is non-nil iff there's an underlying connection attached;
+	// calling it should eventually result in the connection becoming detached,
+	// signaled by the field becoming nil.
+	attached func()
+}
+
+var _ net.Conn = (*Conn)(nil)
+
+const handshakeTimeout = 5 * time.Second
+
+const (
+	errorTag        = ^uint64(0)
+	errorTagUvarint = "\xff\xff\xff\xff\xff\xff\xff\xff\xff\x01"
+)
+
+const maxFrameSize = 128 * 1024
+
+// runResumeV1Unlocking runs the symmetric resumption v1 protocol for r, using
+// nc as the underlying transport. The previous attached transport, if any, will
+// be detached immediately. firstConn signifies that the connection has not been
+// used, and the initial handshake will be assumed to be 0 for both sides. The
+// connection lock is assumed to be held when entering the function, since the
+// correct behavior of firstConn requires no possible external interference
+// before the attach point is reached; the lock will be not held when the
+// function returns.
+func runResumeV1Unlocking(r *Conn, nc net.Conn, firstConn bool) error {
+	defer nc.Close()
+
+	if !firstConn {
+		for !r.remoteClosed && r.attached != nil {
+			r.attached()
+			r.cond.Wait()
+		}
```

Comment:
I don't think it makes too much sense to do it like that because wakeups can happen for all sorts of reasons - a goroutine reading one byte at a time `n` times will generate `n` wakeups and trigger the warning even though there's nothing wrong with the detaching procedure - and counting how many different instances of `requestDetach` the detach loop has encountered is really awkward because `func()` is not comparable (except to `nil`).

Do you have ideas on how to do that?

Reply:
Added a time-based check to the detach loop; we're going to notice if a deadlock completely breaks SSH, but some change that makes detaching really slow will only result in hiccups, so logging that seems appropriate. We do the same thing for slow TLS handshakes and slow sqlite transactions.